File: scripts/combine_mycsv_years.py
#!/usr/bin/env python3
"""Combine year-suffixed CSV exports in mycsv/ into single files.

In `mycsv/`, many tables are exported as one CSV per season/year, e.g.:

  mycsv/f1_official_driver_standings/
    f1_official_driver_standings2023.csv
    f1_official_driver_standings2024.csv
    f1_official_driver_standings2025.csv

This script will concatenate those into:

  mycsv/f1_official_driver_standings/f1_official_driver_standings.csv

It keeps the existing per-year files (safe) and writes the combined file.

Usage:
  python scripts/combine_mycsv_years.py
  python scripts/combine_mycsv_years.py --dry-run
  python scripts/combine_mycsv_years.py --table f1_official_driver_standings

Notes:
- Assumes all year files for a table share the same header.
- Will refuse to combine if headers differ.
"""

from __future__ import annotations

import argparse
import csv
import re
from pathlib import Path
from typing import Iterable

YEAR_SUFFIX_RE = re.compile(r"^(?P<base>.+?)(?P<year>20\d\d)\.csv$", re.IGNORECASE)


def iter_year_files(table_dir: Path) -> list[Path]:
    files: list[tuple[int, Path]] = []
    for p in table_dir.glob("*.csv"):
        m = YEAR_SUFFIX_RE.match(p.name)
        if not m:
            continue
        year = int(m.group("year"))
        files.append((year, p))
    files.sort(key=lambda t: t[0])
    return [p for _, p in files]


def read_header(path: Path) -> list[str]:
    # Handle possible UTF-8 BOM.
    with path.open("r", encoding="utf-8-sig", newline="") as f:
        reader = csv.reader(f)
        return next(reader)


def write_combined(out_path: Path, inputs: Iterable[Path], *, dry_run: bool) -> int:
    inputs = list(inputs)
    if not inputs:
        return 0

    header0 = read_header(inputs[0])
    for p in inputs[1:]:
        h = read_header(p)
        if h != header0:
            raise SystemExit(
                "Header mismatch while combining:\n"
                f"  - {inputs[0].name}: {header0}\n"
                f"  - {p.name}: {h}\n"
                "Fix the exports or keep them separate."
            )

    if dry_run:
        print(f"[dry-run] Would write {out_path} from {len(inputs)} files")
        for p in inputs:
            print(f"  - {p.name}")
        return 0

    out_path.parent.mkdir(parents=True, exist_ok=True)
    with out_path.open("w", encoding="utf-8", newline="") as out_f:
        writer = csv.writer(out_f)
        writer.writerow(header0)

        for p in inputs:
            with p.open("r", encoding="utf-8-sig", newline="") as in_f:
                reader = csv.reader(in_f)
                # skip header
                try:
                    next(reader)
                except StopIteration:
                    continue
                for row in reader:
                    writer.writerow(row)

    print(f"Wrote {out_path} ({len(inputs)} inputs)")
    return 0


def main() -> int:
    ap = argparse.ArgumentParser()
    ap.add_argument("--table", help="Optional: only combine one table folder under mycsv/")
    ap.add_argument("--dry-run", action="store_true")
    args = ap.parse_args()

    repo_root = Path(__file__).resolve().parents[1]
    mycsv_dir = repo_root / "mycsv"

    if not mycsv_dir.exists():
        raise SystemExit(f"mycsv/ not found at: {mycsv_dir}")

    table_dirs: list[Path]
    if args.table:
        table_dirs = [mycsv_dir / args.table]
    else:
        table_dirs = [p for p in mycsv_dir.iterdir() if p.is_dir()]

    for table_dir in sorted(table_dirs):
        if not table_dir.exists():
            raise SystemExit(f"Table folder not found: {table_dir}")

        year_files = iter_year_files(table_dir)
        if not year_files:
            continue

        # Determine base name from first file
        m = YEAR_SUFFIX_RE.match(year_files[0].name)
        assert m
        base = m.group("base")
        out_path = table_dir / f"{base}.csv"

        write_combined(out_path, year_files, dry_run=args.dry_run)

    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/combine_mycsv_years.mjs
#!/usr/bin/env node
/**
 * Combine year-suffixed CSV exports in mycsv/ into single files.
 *
 * Example input:
 *   mycsv/f1_official_driver_standings/f1_official_driver_standings2023.csv
 *   mycsv/f1_official_driver_standings/f1_official_driver_standings2024.csv
 *   mycsv/f1_official_driver_standings/f1_official_driver_standings2025.csv
 *
 * Output:
 *   mycsv/f1_official_driver_standings/f1_official_driver_standings.csv
 *
 * Keeps per-year files and writes/overwrites the combined file.
 *
 * Usage:
 *   node scripts/combine_mycsv_years.mjs
 *   node scripts/combine_mycsv_years.mjs --dry-run
 *   node scripts/combine_mycsv_years.mjs --table f1_official_driver_standings
 */

import fs from 'node:fs/promises';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const YEAR_SUFFIX_RE = /^(?<base>.+?)(?<year>20\d\d)\.csv$/i;

function parseArgs(argv) {
  const out = { dryRun: false, table: null };
  for (let i = 2; i < argv.length; i++) {
    const a = argv[i];
    if (a === '--dry-run') out.dryRun = true;
    else if (a === '--table') out.table = argv[++i];
  }
  return out;
}

function stripBom(s) {
  return s.charCodeAt(0) === 0xfeff ? s.slice(1) : s;
}

function splitLines(s) {
  // We assume these CSVs don't contain embedded newlines inside quoted fields.
  // (true for our generated exports)
  return s.replace(/\r\n/g, '\n').replace(/\r/g, '\n').split('\n');
}

async function listYearFiles(tableDir) {
  const entries = await fs.readdir(tableDir, { withFileTypes: true });
  const files = [];
  for (const e of entries) {
    if (!e.isFile()) continue;
    if (!e.name.toLowerCase().endsWith('.csv')) continue;
    const m = e.name.match(YEAR_SUFFIX_RE);
    if (!m) continue;
    files.push({ year: Number(m.groups.year), name: e.name, base: m.groups.base });
  }
  files.sort((a, b) => a.year - b.year);
  return files;
}

async function combineTable(tableDir, files, { dryRun }) {
  const base = files[0].base;
  const outPath = path.join(tableDir, `${base}.csv`);

  const contents = [];
  for (const f of files) {
    const p = path.join(tableDir, f.name);
    const raw = await fs.readFile(p, 'utf8');
    const text = stripBom(raw);
    const lines = splitLines(text).filter((l) => l.length > 0);
    if (lines.length === 0) continue;
    contents.push({ name: f.name, header: lines[0], rows: lines.slice(1) });
  }

  if (contents.length === 0) return;

  const header0 = contents[0].header;
  for (const c of contents.slice(1)) {
    if (c.header !== header0) {
      throw new Error(
        `Header mismatch while combining ${path.basename(tableDir)}\n` +
          `  - ${contents[0].name}: ${header0}\n` +
          `  - ${c.name}: ${c.header}`
      );
    }
  }

  const outLines = [header0];
  for (const c of contents) outLines.push(...c.rows);
  const outText = outLines.join('\n') + '\n';

  if (dryRun) {
    console.log(`[dry-run] Would write ${outPath} from ${contents.length} files`);
    for (const c of contents) console.log(`  - ${c.name}`);
    return;
  }

  await fs.writeFile(outPath, outText, 'utf8');
  console.log(`Wrote ${outPath} (${contents.length} inputs)`);
}

async function main() {
  const args = parseArgs(process.argv);
  const __filename = fileURLToPath(import.meta.url);
  const __dirname = path.dirname(__filename);
  const repoRoot = path.resolve(__dirname, '..');
  const mycsvDir = path.join(repoRoot, 'mycsv');

  const tableDirs = [];
  if (args.table) {
    tableDirs.push(path.join(mycsvDir, args.table));
  } else {
    const entries = await fs.readdir(mycsvDir, { withFileTypes: true });
    for (const e of entries) {
      if (e.isDirectory()) tableDirs.push(path.join(mycsvDir, e.name));
    }
  }

  tableDirs.sort();
  for (const td of tableDirs) {
    const files = await listYearFiles(td);
    if (files.length === 0) continue;
    await combineTable(td, files, { dryRun: args.dryRun });
  }
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});
